src/main/java/com/github/manolo8/darkbot/gui/titlebar/TrayButton.java

```diff
@@ -1,30 +1,119 @@
 
     private final TrayIcon icon;
+    // Default PopupMenu for TrayIcon has no look & feel, forcing us to use a JPopupMenu
+    private final JPopupMenu popupMenu;
+    // We use a JDialog to make JPopupMenu disappear when it loses focus, see https://stackoverflow.com/a/20079304
+    private final JDialog dialog;
+
     private boolean shownMsg;
+    private Main main;
 
-    TrayButton(JFrame frame) {
+    TrayButton(Main main, JFrame frame) {
         super(UIUtils.getIcon("tray"), frame);
         setToolTipText(I18n.get("gui.tray_button"));
         setVisible(SystemTray.isSupported());
-        icon = new TrayIcon(MainGui.ICON, "DarkBot");
+
+        this.main = main;
+        popupMenu = createPopup();
+        dialog = createDialog();
+        icon = createTrayIcon();
+    }
+
+    private TrayIcon createTrayIcon() {
+        TrayIcon icon = new TrayIcon(MainGui.ICON, "DarkBot");
         icon.setImageAutoSize(true);
+
         icon.addActionListener(l -> {
             SystemTray.getSystemTray().remove(icon);
             frame.setVisible(true);
+            super.setVisible(true);
+        });
+
+        icon.addMouseListener(new MouseAdapter() {
+            @Override
+            public void mouseClicked(MouseEvent e) {
+                if (SwingUtilities.isRightMouseButton(e)) {
+                    updatePopupMenu();
+                    popupMenu.setLocation(e.getX(), e.getY() - popupMenu.getPreferredSize().height);
+                    popupMenu.setInvoker(popupMenu);
+                    dialog.setLocation(e.getX(), e.getY());
+
+                    dialog.setVisible(true);
+                    popupMenu.setVisible(true);
+                }
+            }
+        });
+        return icon;
+    }
+
+    private JPopupMenu createPopup() {
+        JPopupMenu popup = new JPopupMenu("DarkBot");
+
+        JMenuItem title = new JMenuItem("DarkBot", UIUtils.getIcon("icon"));
+        JMenuItem quit = new JMenuItem(I18n.get("gui.tray_menu.quit"));
+
+        title.setEnabled(false);
+        title.setDisabledIcon(UIUtils.getIcon("icon"));
+        quit.addActionListener(l -> {
+            System.out.println("Tray icon exit button pressed, exiting");
+            System.exit(0);
+        });
+
+        popup.add(title);
+        popup.add(new JPopupMenu.Separator());
+        new ExtraButton.DefaultExtraMenuProvider()
+                .getExtraMenuItems(main).forEach(popup::add);
+        popup.add(new JPopupMenu.Separator());
+        popup.add(quit);
+
+        return popup;
+    }
+
+    private void updatePopupMenu() {
+        List<JComponent> defaultExtraMenu = (List<JComponent>) new ExtraButton.DefaultExtraMenuProvider()
+                .getExtraMenuItems(main);
+
+        while (popupMenu.getComponentCount() > 4)
+            popupMenu.remove(2);
+
+        Collections.reverse(defaultExtraMenu);
+        for (JComponent component : defaultExtraMenu)
+            popupMenu.add(component, 2);
+    }
```

Comment:
This is quite an overcomplicated way to update the menu, just remove all entires and re-add them, use the same logic from createPopup to add the entires